### PR TITLE
Working driver metadata storage and retrieval.

### PR DESCRIPTION
**Pull request by petewa**
### Working driver metadata storage and retrieval.

Storage of driver metadata during ingest and loading via BandInfo was partially implemented.
This PR addresses this.

### Proposed changes

- Store metadata from `write_dataset_to_storage` in `ingest.py`
- Load into `BandInfo.driver_data` and made available to the Reader Driver during a `dc.load`

 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/storage/_base.py [hunk 1]
```diff
@@ -42,11 +42,10 @@ def _get_band_and_layer(b: Dict[str, Any]) -> Tuple[Optional[int], Optional[str]
 
 
 def _extract_driver_data(ds: Dataset) -> Optional[Any]:
-    dd = getattr(ds, 'driver_data', None)
-    if dd is None:
-        dd = getattr(ds, 's3_metadata', None)  # TODO: change CSIRO driver to populate `driver_data`
-    return dd
-
+    if 'storage_metadata' in ds.metadata_doc:
```
Comment (Kirill888) [inline datacube/storage/_base.py]: 1. Let's keep `driver_data` key, I think `driver data` communicates more clearly that this data is not interpreted by datacube itself and is only passed on to the driver for interpretation.

2. Can you also please change it to `return ds.metadata_doc.get('driver_data', None)` to reduce need for testing branches, it is also a safer paradigm for multi-threaded applications.

File: datacube/scripts/ingest.py [hunk 2]
```diff
@@ -304,6 +304,8 @@ def _index_datasets(index, results):
             extra_args['storage_metadata'] = datasets.attrs['storage_metadata']
 
         for dataset in datasets.values:
+            if 'storage_metadata' in extra_args:
```
Comment (Kirill888) [inline datacube/scripts/ingest.py]: Let's call it  `driver_data` see my next comment for justification. The whole dataset document is a kind of "storage metadata", so let's be more specific here.

**Review by Kirill888 — CHANGES_REQUESTED** [review]
Mainly I would like renaming `storage_metadata` -> `driver_data`. 

I'm also not sure about reaching directly into `metadata_doc` without using "metadata indirection mechanism", but I don't like metadata indirection either, so ... it's fine.

Updating some docs would be nice also, maybe in the io driver section.

File: datacube/storage/_base.py (hunk 1 above)
Comment (petewa) [reply datacube/storage/_base.py]: Done

File: datacube/scripts/ingest.py (hunk 2 above)
Comment (petewa) [reply datacube/scripts/ingest.py]: Done

**Review by Kirill888 — APPROVED** [review]
thanks

**Pull request merged**